endpoints/setuid.go

```diff
@@ -148,6 +168,135 @@ func NewSetUIDEndpoint(cfg *config.Configuration, syncersByBidder map[string]use
 	})
 }
 
+// extractGDPRInfo looks for the GDPR consent string and GDPR signal in the GPP query params
+// first and the 'gdpr' and 'gdpr_consent' query params second. If found in both, throws a
+// warning. Can also throw a parsing or validation error
+func extractGDPRInfo(query url.Values) (reqInfo gdpr.RequestInfo, err error) {
+
+	reqInfo, err = parseGDPRFromGPP(query)
+	if err != nil {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, err
+	}
+
+	legacySignal, legacyConsent, err := parseLegacyGDPRFields(query, reqInfo.GDPRSignal, reqInfo.Consent)
+	isWarning := errortypes.IsWarning(err)
+
+	if err != nil && !isWarning {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, err
+	}
+
+	// If no GDPR data in the GPP fields, use legacy instead
+	if reqInfo.Consent == "" && reqInfo.GDPRSignal == gdpr.SignalAmbiguous {
+		reqInfo.GDPRSignal = legacySignal
+		reqInfo.Consent = legacyConsent
+	}
+
+	if reqInfo.Consent == "" && reqInfo.GDPRSignal == gdpr.SignalYes {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, errors.New("GDPR consent is required when gdpr signal equals 1")
+	}
+
+	return reqInfo, err
+}
+
+// parseGDPRFromGPP parses and validates the "gpp_sid" and "gpp" query fields.
+func parseGDPRFromGPP(query url.Values) (gdpr.RequestInfo, error) {
+	var gdprSignal gdpr.Signal = gdpr.SignalAmbiguous
+	var gdprConsent string = ""
+	var err error
+
+	gdprSignal, err = parseSignalFromGppSidStr(query.Get("gpp_sid"))
+	if err != nil {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, err
+	}
+
+	gdprConsent, err = parseConsentFromGppStr(query.Get("gpp"))
+	if err != nil {
+		return gdpr.RequestInfo{GDPRSignal: gdpr.SignalAmbiguous}, err
+	}
+
+	return gdpr.RequestInfo{
+		Consent:    gdprConsent,
+		GDPRSignal: gdprSignal,
+	}, nil
+}
+
+// parseLegacyGDPRFields parses and validates the "gdpr" and "gdpr_consent" query fields which
+// are considered deprecated in favor of the "gpp" and "gpp_sid". The parsed and validated GDPR
+// values contained in "gpp" and "gpp_sid" are passed in the parameters gppGDPRSignal and
+// gppGDPRConsent. If the GPP parameters come with non-default values, this function discards
+// "gdpr" and "gdpr_consent" and returns a warning.
+func parseLegacyGDPRFields(query url.Values, gppGDPRSignal gdpr.Signal, gppGDPRConsent string) (gdpr.Signal, string, error) {
+	var gdprSignal gdpr.Signal = gdpr.SignalAmbiguous
+	var gdprConsent string
+	var warning error
+
+	if gdprQuerySignal := query.Get("gdpr"); len(gdprQuerySignal) > 0 {
+		if gppGDPRSignal == gdpr.SignalAmbiguous {
+			if gdprQuerySignal != "" && gdprQuerySignal != "0" && gdprQuerySignal != "1" {
+				return gdpr.SignalAmbiguous, "", errors.New("the gdpr query param must be either 0 or 1. You gave " + gdprQuerySignal)
+			}
+
+			if i, err := strconv.Atoi(gdprQuerySignal); err == nil {
```

Comment:
Since there are only 2 possible values, would a case statement be more efficient? This could even be combined with the above check, so if it does not match one of the two values, the default case is the above error.

Reply:
Makes perfect sense. Modified